### PR TITLE
Collection super class

### PR DESCRIPTION
**Pull request by iEv0lv3**
## Collection Super Class

An absolute monster of a pull-request that spans most files in the project. The main goal of this PR is to update our project to use inheritance by implementing a `Collection.rb` super class. The `Collection` super class the has sub-classes of `GameCollection`, `TeamCollection`, `GameTeamCollection`, and `SeasonCollection`. 

* Update naming conventions across classes for instance variables, parameters, and arguments
* Update collection classes that existed previously to use inheritance from the new `Collection` super class
* Update all `stat_tracker_test.rb` tests to work with `Collection.rb` super class
* Update all  `stat_tracker_test.rb` tests and setup to exactly reflect the requirements in rspec harness
* Update some `stat_tracker_test.rb` tests to use mocks and stubs
* Remove unnecessary and duplicate files
* Update all methods in `stat_tracker.rb` to work with new `Collection.rb` super class collections
* Create `season.rb` class to group games into individual seasons
* Create `season_collection.rb` class which contains a hash of multiple seasons of game data, with the season years being the key, and all `Game` objects being added to a hash as the value.
* Update `game.rb`, `team.rb`,  and `game_team.rb` files to use `(row)` as the initialize parameter
* Update csv data file headers to properly reflect Ruby syntax
  * This also properly imports those data types to the collections now
* Test coverage is currently at 96%

## To Do
* Tests for `game.rb`, `team.rb`, `game_team.rb`, and their collections `game_collection`, `team_collection`, and `game_team_collection` should be updated individually. All the classes are functioning properly when using `stat_tracker.rb` and are covered by tests there. However they have not been re-written individually after the refactor.
* After confirming functionality and merging, the `dev` branch should be merged into `master`

### CODE REVIEW

File: lib/season_collection.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+class SeasonCollection < Collection
+  def initialize(csv_file_path)
+    super(csv_file_path, Season)
+  end
+
+  def create_collection(csv_file_path, collection_type)
+    from_csv(csv_file_path).reduce(Hash.new([])) do |hash, row|
+      hash[row[:season]] += [collection_type.new(row)]
+      hash
+    end
```
Comment (alerrian) [inline lib/season_collection.rb:13]: This method is replacing the Collection create_collection method due to slight difference in how this collection is being created. Nice use of inheritance!

File: lib/stat_tracker.rb [hunk 2]
```diff
@@ -1,41 +1,45 @@
...
 
-  def initialize(game_file, team_file)
-    @games_collection = GameCollection.new(game_file)
-    @teams_collection = TeamCollection.new(team_file)
-    @game_teams_collection = GameTeamCollection.new(game_team_file)
+  def initialize(games, teams, game_teams)
+    @games_collection = GameCollection.new(games)
+    @teams_collection = TeamCollection.new(teams)
+    @game_teams_collection = GameTeamsCollection.new(game_teams)
   end
```
Comment (alerrian) [inline lib/stat_tracker.rb:31]: Just want to make sure we actually add an initialization for seasons here so we can actually use it!

File: lib/team_collection.rb [hunk 3]
```diff
@@ -1,20 +1,8 @@
...
-    CSV.read(csv_file_path, headers: true, header_converters: :symbol)
-  end
-
-  def create_teams(csv_file_path)
-    from_csv(csv_file_path).map do |row|
-      Team.new(row)
-    end
+    super(csv_file_path, Team)
   end
 end
```
Comment (alerrian) [inline lib/team_collection.rb:8]: Love how tiny these classes are now due to an awesome inheritance tree.

File: test/collection_test.rb [hunk 4]
```diff
@@ -1,12 +1,12 @@
-require 'simplecov'
-SimpleCov.start
-require 'minitest/autorun'
-require 'minitest/pride'
+require_relative './test_helper'
 require 'csv'
 require './lib/collection'
 require './lib/game'
 require './lib/team'
-require './lib/game_team'
```
Comment (alerrian) [inline test/collection_test.rb:9]: Good call on fixing the names here. Makes it a little easier to understand when we see them in the wild.

File: test/season_test.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    @season.stubs(:game_id).returns('2012030233')
+    @season.stubs(:type).returns('Postseason')
+    @season.stubs(:date_time).returns('5/20/13')
+    @season.stubs(:home_team_id).returns(17)
+    @season.stubs(:away_team_id).returns(16)
+    @season.stubs(:away_goals).returns(1)
+    @season.stubs(:home_goals).returns(3)
+    @season.stubs(:venue).returns('Dignity Health Sports Park')
+    @season.stubs(:venue_link).returns('/api/v1/venues/null')
+  end
```
Comment (alerrian) [inline test/season_test.rb:19]: Excellent use of stubs here. I think stubbing the names of the identifiers really hooks us up for test performance and allows us to bypass a lot of running of the csv files over and over.

File: test/stat_tracker_test.rb [hunk 6]
```diff
@@ -3,98 +3,113 @@
...
-    games = GameCollection.new(game_path)
-    teams = TeamCollection.new(team_path)
-    # game_team = GameTeamCollection.new(game_team_path)
-    @new_tracker = StatTracker.new(games, teams)
+    game_path = './data/games.csv'
+    team_path = './data/teams.csv'
+    game_teams_path = './data/game_teams.csv'
+
+    locations = {
+                  games: game_path,
```
[the remark this answers is not in the document]
Comment (alerrian) [reply test/stat_tracker_test.rb]: I would say that this Hound style is one you should look at following. I am pretty certain the style guide dictates hashes that are listed across several lines should be listed tab-indented (2-spaces) from the leftmost position of where the hash is declared. [Style Guide](https://github.com/rubocop-hq/ruby-style-guide#no-double-indent)

File: test/stat_tracker_test.rb [hunk 7]
```diff
@@ -3,98 +3,113 @@
...
 
   def test_stat_tracker_average_goals_by_season
     average_hash = {
-      "20122013"=>3.85, 
-      "20142015"=>4.02, 
-      "20152016"=>4.1, 
-      "20162017"=>4.36, 
-      "20172018"=>4.16, 
-      "20132014"=>4.27
+      "20122013"=>4.12,
```
[the remark this answers is not in the document]
Comment (alerrian) [reply test/stat_tracker_test.rb]: I think we can and should ignore this one.

File: test/stat_tracker_test.rb [hunk 8]
```diff
@@ -3,98 +3,113 @@
...
+      "20132014"=>4.19,
+      "20142015"=>4.14,
+      "20152016"=>4.16,
+      "20162017"=>4.23,
+      "20172018"=>4.44,
     }
 
-    assert_equal average_hash, @new_tracker.average_goals_by_season
```
Comment (alerrian) [inline test/stat_tracker_test.rb:47]: Good call on these names. stat_tracker makes much MUCH more sense.

**Review by alerrian — COMMENTED** [review]
Excellent use of stubs for testing parameters. Overall, a great update to our existing code base that adds inheritance, several new methods and updates pointers and methods to accommodate better structure overall.

File: lib/stat_tracker.rb [hunk 9]
```diff
@@ -1,41 +1,51 @@
...
 
-  def initialize(game_file, team_file)
-    @games_collection = GameCollection.new(game_file)
-    @teams_collection = TeamCollection.new(team_file)
-    @game_teams_collection = GameTeamCollection.new(game_team_file)
+  def initialize(games, teams, game_teams)
+    @game_collection = GameCollection.new(games)
+    @team_collection = TeamCollection.new(teams)
+    @season_collection = SeasonCollection.new(games)
```
Comment (alerrian) [inline lib/stat_tracker.rb:29]: This was the change for this file. Looks great. Good to see it initialized here so we can begin using it.

File: test/stat_tracker_test.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (iEv0lv3) [reply test/stat_tracker_test.rb]: I agree, and think I have a misunderstanding of how this works. Currently I tabbed in from the start of the hash curly brace, but that didn't seem to do it. What am I missing?

File: lib/season_collection.rb (hunk 1 above)
Comment (iEv0lv3) [reply lib/season_collection.rb:13]: Thanks for noticing! Pretty stoked about that one 😄

File: test/stat_tracker_test.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (alerrian) [reply test/stat_tracker_test.rb]: You should write the `({` then hit return, tab in(if it isn't automatic) then write one key-val pair per line. `})` should be inline with the name of the hash.

File: lib/stat_tracker.rb (hunk 9 above)
Comment (iEv0lv3) [reply lib/stat_tracker.rb:29]: Nice! First time I've updated a PR in progress; great to know how seamless the process is.

**Comment by alerrian** [comment]
All-in-all a great bundle of changes that adds inheritance, some style fixes and better readability to our entire infrastructure. Great work. Keep an eye on some of those style fixes that will plague us for all time!

**Pull request merged**